### PR TITLE
Handle interruption in RetryDriver and TaskRunner

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description
This PR includes two changes to Thread interrupt handling:
1. In `trino-hive`, modify `RetryDriver` to stop attempting retries if `InterruptedException` is caught or the current thread is interrupted and immediately throw the current exception. Otherwise, `RetryDriver` might interfere with drivers terminating in a timely manner after tasks are terminated via a cancel, failure, or abort.
2. In `trino-main`, check and clear thread interrupt status at the end of processing a split in `TaskRunner` before looping back to start processing a new split. Otherwise, a split that is interrupted (eg: because the task was cancelled) would leave the thread interrupted flag set which would cause the `TaskRunner` to stop processing and unnecessarily submit a new instance from which could cause additional threads to be created in the cached thread pool executor.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java [hunk 1]
```diff
@@ -561,6 +561,15 @@ public void run()
                         }
                         splitFinished(split);
                     }
+                    finally {
+                        // Clear the interrupted flag on the current thread, driver cancellation may have triggered an interrupt
+                        if (Thread.interrupted()) {
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:566]: What is the motivation of this change?

Based on my understanding terminating a runner tread upon interruption is by design (see line 503, `while (!closed && !Thread.currentThread().isInterrupted()) {`)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/RetryDriver.java [hunk 2]
```diff
@@ -130,6 +130,11 @@ public <V> V run(String callableName, Callable<V> callable)
                 return callable.call();
             }
             catch (Exception e) {
+                // Immediately stop retry attempts once an interrupt has been received
+                if (e instanceof InterruptedException || Thread.currentThread().isInterrupted()) {
```
Comment (arhimondr) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/RetryDriver.java:134]: I think we should recover the `Interrupted` flag while also stopping retries.

File: core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java (hunk 1 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:566]: Terminating a runner isn't necessary if the `TaskExecutor` is not closed. Essentially there are two sources of interrupts that can occur:
1. `TaskExecutor#shutdown()` sets `closed = true`, calls `executor.shutdownNow()` which will interrupt all threads. If they're blocked in a call to `waitingSplits.take()`, they will wake up and throw `InterruptedException` and the runner will terminate (without adding a replacement for itself). This happens exactly once during the whole lifecycle of the JVM in production, but in tests obviously it can happen multiple times.
2. The active driver is closed while the `TaskRunner` is running it. The task runner thread is interrupted and may or may not throw an `InterruptedException` as a result since it's in the middle of a call to `split.process()` and we can't know for sure that an interruptable method will be called. Alternatively, some code inside of the driver interrupts the current thread for some other reason. Regardless of when/why that happens, the runner will terminate on the next iteration if we leave the flag set (because of `!Thread.currentThread().isInterrupted()`) but if `TaskExecutor#shutdown()` was _not_ called then there's no need to let that happen since it will just re-queue a new `TaskRunner` to replace itself via line 576 in the finally block.

So we can see that there are essentially two points to interrupting the thread: 1. stop all `TaskRunner`s because we're shutting down (rare) or 2. stop a single driver that happens to be running on a given `TaskRunner`. In the second case, pushing a new runner to replace the one that will terminate into the executor will create a new thread (because it's a cached threadpool) if an extra thread doesn't already exist there, which isn't necessary and does have a real runtime memory and performance cost.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/RetryDriver.java (hunk 2 above)
Comment (pettyjamesm) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/RetryDriver.java:134]: I was on the fence about whether to do that here. I think since we're re-throwing `InterruptedException` directly, the caller is supposed to be responsible for making that decision- otherwise you can end up setting an "extra" interrupts on the current thread (perhaps inappropriately?).

Comment (arhimondr) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/RetryDriver.java:134]: Nevermind, I think I misread. You are re-throwing the `InterruptedException` and not suppressing it. No need to recover the flag.

**Review by arhimondr — APPROVED** [review]
